### PR TITLE
Remove the use of Jackson when processing patches

### PR DESCRIPTION
**Pull request by bdemers**
- FitlerExpressions.inMemory() now returns Predicate<Object> instead of KeyedResource
- Added test for complex patch filter

### CODE REVIEW

File: scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java [hunk 1]
```diff
@@ -104,28 +92,19 @@ private <T extends ScimResource> T apply(final T source, final PatchOperation pa
     if (valuePathExpression.getAttributeExpression() != null && attributeObject instanceof Collection<?>) {
       // apply expression filter
       Collection<Object> items = (Collection<Object>) attributeObject;
-      Collection<Object> updatedCollection = items.stream().map(item -> {
-        KeyedResource keyedResource = (KeyedResource) item;
```
Comment (bdemers) [inline scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java:108]: @erant10 I think this was a `KeyedResource` because it was a common parent, but nothing is requiring these objects to be that type, so I changed that method to return a `Predicate<Object>` instead.

File: scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java [hunk 2]
```diff
@@ -104,28 +92,19 @@ private <T extends ScimResource> T apply(final T source, final PatchOperation pa
...
-          if (keyedResourceAsMap.get(subAttributeName) == null && patchOperation.getOperation().equals(PatchOperation.Type.REPLACE)) {
-            throw new IllegalArgumentException("Cannot apply patch replace on missing property: " + valuePathExpression.toFilter());
-          }
-          keyedResourceAsMap.put(subAttributeName, patchOperation.getValue());
-          return keyedResourceAsMap;
-        } else {
-          // filter does not apply
-          return item;
+          Schema.AttributeAccessor subAttributeAccessor = attribute.getAttribute(subAttributeName).getAccessor();
+          subAttributeAccessor.set(item, patchOperation.getValue());
```
Comment (bdemers) [inline scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java]: I'm not 100% sure this will fit what you had in mind originally, but we should be able to use the accessor `set` method here.
We may need to test cases where the field is immutable, (and make sure the correct exception is thrown), but that _should_ happen at the accessor level (and ideally should happen with the ObjectMapper too).

I'll try to create a test for this (as that's independent of my change here)

File: scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java [hunk 3]
```diff
@@ -104,28 +92,19 @@ private <T extends ScimResource> T apply(final T source, final PatchOperation pa
...
       Collection<Object> items = (Collection<Object>) attributeObject;
-      Collection<Object> updatedCollection = items.stream().map(item -> {
-        KeyedResource keyedResource = (KeyedResource) item;
-        Map<String, Object> keyedResourceAsMap = objectAsMap(item);
-        Predicate<KeyedResource> pred = FilterExpressions.inMemory(valuePathExpression.getAttributeExpression(), baseSchema);
-        if (pred.test(keyedResource)) {
+      items.forEach(item -> {
+        Predicate<Object> pred = FilterExpressions.inMemory(valuePathExpression.getAttributeExpression(), baseSchema);
+        if (pred.test(item)) {
           String subAttributeName = valuePathExpression.getAttributePath().getSubAttributeName();
```
Comment (bdemers) [inline scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java]: One thing we may need to consider with either of these solutions is that they assume a single nested value. `email.types`, `address.city`, etc.

IIRC, all of the official spec schemas only have one level like this, but it's possible there could be a more complex value in the future or through a custom extension.

That said, I'm mostly just adding a note to think about it; I don't think it needs to block this getting merged (the `develop` branch just throws an exception, so this is a huge improvement and probably hits 90% of cases)

**Comment by bdemers** [comment]
In hacking on this a bit more, I think your original use of Jackson is probably the better path.

My initial pass at a more complex test, made some bad assumptions.

I was assuming the `value` of the Patch Op was already marshaled. This isn't the case, and it could be more complex than a single object, for example:

```json
   {
     "schemas":
       ["urn:ietf:params:scim:api:messages:2.0:PatchOp"],
     "Operations":[{
       "op":"add",
       "value":{
         "emails":[
           {
             "value":"babs@jensen.org",
             "type":"home"
           }
         ],
         "nickname":"Babs"
     }]
   }
```

In this case, both the `nickname` value must be set and the new email added.
I was already bumping into issues with trying to figure out the correct collection type to use when new elements were added.

TL;DR, me trying to avoid the extra round trip with Jackson creates a ton more work.

Sorry for the confusion around this.
